### PR TITLE
db/hints: Move the code for writing hints to a separate function

### PR DESCRIPTION
**Pull request by dawmd**
In scylladb/scylladb@7301a96, in the function `hint_endpoint_manager::store_hint()`,
we transformed the lambda passed to `seastar::with_gate()` to a coroutine lambda
to improve the readability. However, there was a subtle problem related to
lifetimes of the captures that needed to be addressed:

* Since we started `co_await`ing in the lambda, the captures were at risk of
  being destructed too soon. The usual solution is to wrap a coroutine lambda
  within a `seastar::coroutine::lambda` object and rely on the extended lifetime
  enforced by the semantics of the language.
  See `docs/dev/lambda-coroutine-fiasco.md` for more context.
* However, since we don't immediately `co_await` the future returned by
  `with_gate()`, we cannot rely on the extended lifetime provided by the wrapper.
  The document linked in the previous bullet point suggests keeping the passed
  coroutine lambda as a variable and pass it as a reference to `with_gate()`.
  However, that's not feasible either because we discard the returned future and
  the function returns almost instantly -- destructing every local object, which
  would encompass the lambda too.

The solution used in the commit was to move captures of the lambda into
the lambda's body. That helped because Seastar's backend is responsible for
keeping all of the local variables alive until the lambda finishes its execution.
However, we didn't move all of the captures into the lambda -- the missing one
was the `this` pointer that was implicitly used in the lambda.

Address sanitiser hasn't reported any bugs related to the pointer yet, but
the bug is most likely there.

In this commit, we transform the lambda's body into a new member function
and only call it from the lambda. This way, we don't need to care about
the lifetimes of the captures because Seastar ensures that the function's
arguments stay alive until the coroutine finishes.

Choosing this solution instead of assigning `this` to a pointer variable
inside the lambda's body and using it to refer to the object's members
has actual benefit: it's not possible to accidentally forget to refer
to a member of the object via the pointer; it also makes the code less
awkward.

Fixes scylladb/scylladb#20306

### CODE REVIEW

File: db/hints/internal/hint_endpoint_manager.cc [hunk 1]
```diff
@@ -53,46 +53,47 @@ bool hint_endpoint_manager::store_hint(schema_ptr s, lw_shared_ptr<const frozen_
...
-            shard_stats().size_of_hints_in_progress += mut_size;
+            self.shard_stats().size_of_hints_in_progress += mut_size;
 
             if (utils::get_local_injector().enter("slow_down_writing_hints")) {
                 co_await seastar::sleep(std::chrono::seconds(10));
             }
 
             try {
-                const auto shared_lock = co_await get_shared_lock(file_update_mutex());
+                const auto shared_lock = co_await get_shared_lock(self.file_update_mutex());
```
Comment (tchaikov) [inline db/hints/internal/hint_endpoint_manager.cc:70]: @dawmd hi Dawid, i don't understand this change. what exactly the "captures" you want to keep alive? if it is `this`, then how can we extend the life cycle of `this` by holding a weak reference to it?

Comment (dawmd) [reply db/hints/internal/hint_endpoint_manager.cc:70]: `this` is a pointer, but it's still a variable and needs to be stored by the lambda. If the lambda object is freed, that memory will be freed as well and using `this`, implicitly or not, might lead to use-after-free. ASAN hasn't reported any issue with that so far, so maybe something more subtle is going on in this code, but that was my reasoning

File: db/hints/internal/hint_endpoint_manager.cc [hunk 2]
```diff
@@ -53,46 +53,47 @@ bool hint_endpoint_manager::store_hint(schema_ptr s, lw_shared_ptr<const frozen_
             // We discard the future to this call to `with_gate()`, so wrapping this lambda
             // within `coroutine::lambda` will NOT work. That's why we need to copy these
             // variables by hand.
+            hint_endpoint_manager& self = *this;
```
Comment (tchaikov) [inline db/hints/internal/hint_endpoint_manager.cc:56]: so we are suffering from https://github.com/scylladb/seastar/blob/master/doc/lambda-coroutine-fiasco.md#alternative-solution-when-lifetime-extension-cannot-be-used here.

have you considered naming the lambda passed to `with_gate()`, and reference it using `std::ref()`, like:
```c++
auto do_store_hint = [this, s = std::move(s), fm = std::move(fm), tr_state = tr_state] () mutable -> future<> {
    // ....
    // could yield
    // references the captured variables
};
try {
    (void) with_gate(_store_gate, std::ref(do_store_hint));
} catch (...) {
    // 
    return false;
}
return true;
```

IMHO, this approach is better in the sense that we don't have to use `self` where `this` is typically used, it's confusing without context.

Comment (dawmd) [reply db/hints/internal/hint_endpoint_manager.cc]: Exactly! The suspicious thing is ASAN doesn't detect it, so either I misunderstand the semantics here, or we're suffering from UB (seems very unlikely). No matter what it is, I think it's better to get that pointer and be safe (the compiler should optimise it anyway).

**Comment by dawmd** [comment]
Version 2:
* The reference `self` changed to a pointer `thiz`.

**Review by tchaikov — APPROVED** [review]
lgtm

**Comment by dawmd** [comment]
I forgot to change the commit's title.

Version 3:
* Updated the commit's and PR's titles

**Review by tchaikov — APPROVED** [review]

**Review by piodul — CHANGES_REQUESTED** [review]
I can still see some remaining implicit references to `this`, including after the first preemption (`_hints_in_progress`, `_last_written_rp`). I propose the following:

- Remove `this` from the capture list,
- Add a `hint_endpoint_manager* thiz` parameter to the lambda,
- Wrap the lambda with `std::bind(..., this)`.

(Lambda's parameters do not suffer from the same lifetime issues as the stuff in the capture list, so this is a neat trick for addressing the coroutine fiasco problem)

This should make sure that we are no longer accessing via a captured `this`. However, I agree with Avi that a better, long-term solution is to extract the lambda to a method.

**Comment by dawmd** [comment]
Version 4:
* Moved the implementation to a separate member function.
* Updated the commit message, its title, as well as the PR's description and title.

**Comment by dawmd** [comment]
No backport is needed because scylladb/scylladb@7301a96 didn't make into branch-6.0 or branch-6.1

File: db/hints/internal/hint_endpoint_manager.cc [hunk 3]
```diff
@@ -45,54 +45,51 @@ constexpr std::chrono::seconds HINT_FILE_WRITE_TIMEOUT = std::chrono::seconds(2)
 
 } // anonymous namespace
 
-bool hint_endpoint_manager::store_hint(schema_ptr s, lw_shared_ptr<const frozen_mutation> fm, tracing::trace_state_ptr tr_state) noexcept {
+future<> hint_endpoint_manager::do_store_hint(schema_ptr s, lw_shared_ptr<const frozen_mutation> fm, tracing::trace_state_ptr tr_state) {
+    ++_hints_in_progress;
```
Comment (piodul) [inline db/hints/internal/hint_endpoint_manager.cc:49]: The diff is quite unreadable due to the indentation change. Please split it into two commits:

- In one commit, move the lambda to a separate function but try to keep the old indentation in order to minimize the diff,
- In the other commit, fix the indentation.

This is what we usually do. Example: https://github.com/scylladb/scylladb/commit/aa378c458e3e10440b01f45dcce74496b7ad7914 + https://github.com/scylladb/scylladb/commit/66292b4baa27d3f7a86f28634be3da8b6a9450a9

**Comment by dawmd** [comment]
Version 5:
* Split the PR into two commits to minimise the diff and make the changes easier to review

**Review by piodul — APPROVED** [review]

**Comment by piodul** [comment]
Queued

**Pull request merged**